File: integration-tests/cpp/delete-contained-entities/src/peer.cpp
/*
 * integration-tests/cpp/delete-contained-entities -- peer. Direct C++ port
 * of c/delete-contained-entities/src/peer.c -- see that file's header
 * comment for the full scenario rationale and
 * docs/design/integration-test-tier.md for the scenario spec.
 *
 * Required stdout markers: "Create topic:" x3, "Create writer for topic:",
 * "Create reader for topic:" x2, "Peer: session disconnected cleanly."
 * Any failure path prints a line starting "FAIL:" and exits nonzero.
 */
#include "session_event.hpp"
#include "zzdds_cpp.hpp"
#include "dcps_impl.hpp"

#include <atomic>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <memory>
#include <vector>
#include <unistd.h>

namespace {

constexpr int SAMPLE_COUNT = 5;
constexpr int RECEIVE_TARGET = 2;
constexpr int READER_READY_TIMEOUT_MS = 10000;
constexpr int RECEIVE_TIMEOUT_MS = 20000;
constexpr int DISCONNECT_TIMEOUT_MS = 30000;
constexpr int POLL_PERIOD_MS = 20;

struct MatchState {
    std::atomic<bool> ever_matched{false};
    std::atomic<int> matched_current_count{0};
    std::atomic<bool> reader_ready{false};
};

class WriterListener : public ::zzdds::DataWriterListenerExBase {
public:
    explicit WriterListener(MatchState *state) : state_(state) {}

    void on_publication_matched(std::shared_ptr<::DDS::DataWriter> /*writer*/,
                                 ::DDS::PublicationMatchedStatus status) override {
        state_->matched_current_count.store(status.current_count);
        if (status.current_count > 0) state_->ever_matched.store(true);
    }

    void on_reliable_reader_ready(::DDS::InstanceHandle_t /*reader_handle*/, bool is_ready) override {
        if (is_ready) state_->reader_ready.store(true);
    }

private:
    MatchState *state_;
};

class ReaderListener : public ::DDS::DataReaderListenerBase {
public:
    explicit ReaderListener(MatchState *state) : state_(state) {}

    void on_subscription_matched(std::shared_ptr<::DDS::DataReader> /*reader*/,
                                  ::DDS::SubscriptionMatchedStatus status) override {
        state_->matched_current_count.store(status.current_count);
        if (status.current_count > 0) state_->ever_matched.store(true);
    }

private:
    MatchState *state_;
};

uint32_t parse_domain(int argc, char **argv) {
    for (int i = 1; i < argc - 1; i++) {
        if (std::strcmp(argv[i], "-d") == 0 || std::strcmp(argv[i], "--domain") == 0) {
            return static_cast<uint32_t>(std::strtoul(argv[i + 1], nullptr, 10));
        }
    }
    return 0;
}

} // namespace

int main(int argc, char **argv) {
    uint32_t domain_id = parse_domain(argc, argv);

    auto factory = zzdds::create_factory();
    if (!factory) {
        std::fprintf(stderr, "FAIL: createFactory() failed\n");
        return 1;
    }

    auto dp = factory->create_participant(domain_id, ::DDS::DomainParticipantQos::default_value(), nullptr, 0);
    if (!dp) {
        std::fprintf(stderr, "FAIL: create_participant() failed on domain %u\n", domain_id);
        return 1;
    }
    auto dp_handle = dp->native_handle();

    if (SessionEventTypeSupport::register_type(dp_handle) != 0) {
        std::fprintf(stderr, "FAIL: register SessionEventTypeSupport failed\n");
        return 1;
    }

    auto out1_topic = dp->create_topic("SessionOut1", "SessionEvent", ::DDS::TopicQos::default_value(), nullptr, 0);
    auto out2_topic = dp->create_topic("SessionOut2", "SessionEvent", ::DDS::TopicQos::default_value(), nullptr, 0);
    auto in_topic = dp->create_topic("SessionIn", "SessionEvent", ::DDS::TopicQos::default_value(), nullptr, 0);
    if (!out1_topic || !out2_topic || !in_topic) {
        std::fprintf(stderr, "FAIL: create_topic() failed\n");
        return 1;
    }
    std::printf("Create topic: SessionOut1\n");
    std::printf("Create topic: SessionOut2\n");
    std::printf("Create topic: SessionIn\n");

    auto pub = dp->create_publisher(::DDS::PublisherQos::default_value(), nullptr, 0);
    auto sub = dp->create_subscriber(::DDS::SubscriberQos::default_value(), nullptr, 0);
    if (!pub || !sub) {
        std::fprintf(stderr, "FAIL: create_publisher/create_subscriber failed\n");
        return 1;
    }

    auto dw_qos = ::DDS::DataWriterQos::default_value();
    dw_qos.reliability.kind = ::DDS::ReliabilityQosPolicyKind::RELIABLE_RELIABILITY_QOS;
    dw_qos.history.kind = ::DDS::HistoryQosPolicyKind::KEEP_ALL_HISTORY_QOS;

    auto in_dw = pub->create_datawriter(in_topic, dw_qos, nullptr, 0);
    if (!in_dw) {
        std::fprintf(stderr, "FAIL: create_datawriter(SessionIn) failed\n");
        return 1;
    }
    std::printf("Create writer for topic: SessionIn\n");

    auto dr_qos = ::DDS::DataReaderQos::default_value();
    dr_qos.reliability.kind = ::DDS::ReliabilityQosPolicyKind::RELIABLE_RELIABILITY_QOS;
    dr_qos.history.kind = ::DDS::HistoryQosPolicyKind::KEEP_ALL_HISTORY_QOS;

    auto zout1_topic = std::static_pointer_cast<::zzdds::TopicImpl>(out1_topic);
    auto out1_desc = zout1_topic->as_topic_description();
    auto out1_dr = sub->create_datareader(out1_desc, dr_qos, nullptr, 0);
    if (!out1_dr) {
        std::fprintf(stderr, "FAIL: create_datareader(SessionOut1) failed\n");
        return 1;
    }
    std::printf("Create reader for topic: SessionOut1\n");

    auto zout2_topic = std::static_pointer_cast<::zzdds::TopicImpl>(out2_topic);
    auto out2_desc = zout2_topic->as_topic_description();
    auto out2_dr = sub->create_datareader(out2_desc, dr_qos, nullptr, 0);
    if (!out2_dr) {
        std::fprintf(stderr, "FAIL: create_datareader(SessionOut2) failed\n");
        return 1;
    }
    std::printf("Create reader for topic: SessionOut2\n");

    MatchState writer_state, out1_state, out2_state;
    auto writer_listener = std::make_shared<WriterListener>(&writer_state);
    auto zin_dw = std::static_pointer_cast<::zzdds::DataWriterImpl>(in_dw);
    if (zin_dw->set_listener_ex(writer_listener, DDS_PUBLICATION_MATCHED_STATUS) != ::DDS::RETCODE_OK) {
        std::fprintf(stderr, "FAIL: set_listener_ex (writer) failed\n");
        return 1;
    }
    auto out1_listener = std::make_shared<ReaderListener>(&out1_state);
    auto out2_listener = std::make_shared<ReaderListener>(&out2_state);
    if (out1_dr->set_listener(out1_listener, DDS_SUBSCRIPTION_MATCHED_STATUS) != ::DDS::RETCODE_OK ||
        out2_dr->set_listener(out2_listener, DDS_SUBSCRIPTION_MATCHED_STATUS) != ::DDS::RETCODE_OK)
    {
        std::fprintf(stderr, "FAIL: set_listener (reader) failed\n");
        return 1;
    }

    SessionEventDataWriter in_writer(in_dw->native_handle());
    SessionEventDataReader out1_reader(out1_dr->native_handle());
    SessionEventDataReader out2_reader(out2_dr->native_handle());

    // Gate on the session's reader actually being registered, not just
    // matched -- see session.cpp's matching comment for why.
    for (int waited_ms = 0; !writer_state.reader_ready.load(); waited_ms += POLL_PERIOD_MS) {
        if (waited_ms >= READER_READY_TIMEOUT_MS) {
            std::fprintf(stderr, "FAIL: no reliable reader became ready within %ds\n", READER_READY_TIMEOUT_MS / 1000);
            return 1;
        }
        usleep(POLL_PERIOD_MS * 1000);
    }

    for (int seq = 0; seq < SAMPLE_COUNT; seq++) {
        ::SessionEvent ev;
        ev.seq = seq;
        if (in_writer.write(ev) != 0) {
            std::fprintf(stderr, "FAIL: write() failed at seq=%d\n", seq);
            return 1;
        }
    }
    std::printf("Peer: wrote %d samples on SessionIn\n", SAMPLE_COUNT);

    int received1 = 0, received2 = 0;
    for (int waited_ms = 0; received1 < RECEIVE_TARGET || received2 < RECEIVE_TARGET; waited_ms += POLL_PERIOD_MS) {
        if (waited_ms >= RECEIVE_TIMEOUT_MS) {
            std::fprintf(stderr, "FAIL: peer did not receive from session within %ds (out1=%d out2=%d)\n",
                          RECEIVE_TIMEOUT_MS / 1000, received1, received2);
            return 1;
        }
        std::vector<::SessionEvent> values(SAMPLE_COUNT);
        std::vector<DDS_SampleInfo> infos(SAMPLE_COUNT);
        int n1 = out1_reader.take_n(values.data(), infos.data(), SAMPLE_COUNT,
                                     ::DDS::ANY_SAMPLE_STATE, ::DDS::ANY_VIEW_STATE, ::DDS::ANY_INSTANCE_STATE);
        for (int i = 0; i < n1; i++) {
            if (infos[i].valid_data) received1++;
        }
        int n2 = out2_reader.take_n(values.data(), infos.data(), SAMPLE_COUNT,
                                     ::DDS::ANY_SAMPLE_STATE, ::DDS::ANY_VIEW_STATE, ::DDS::ANY_INSTANCE_STATE);
        for (int i = 0; i < n2; i++) {
            if (infos[i].valid_data) received2++;
        }
        if (received1 < RECEIVE_TARGET || received2 < RECEIVE_TARGET) usleep(POLL_PERIOD_MS * 1000);
    }
    std::printf("Peer: received from session (out1=%d out2=%d).\n", received1, received2);

    for (int waited_ms = 0;
         !(writer_state.ever_matched.load() && writer_state.matched_current_count.load() == 0 &&
           out1_state.ever_matched.load() && out1_state.matched_current_count.load() == 0 &&
           out2_state.ever_matched.load() && out2_state.matched_current_count.load() == 0);
         waited_ms += POLL_PERIOD_MS)
    {
        if (waited_ms >= DISCONNECT_TIMEOUT_MS) {
            std::fprintf(stderr, "FAIL: peer never saw a clean disconnect within %ds\n", DISCONNECT_TIMEOUT_MS / 1000);
            return 1;
        }
        usleep(POLL_PERIOD_MS * 1000);
    }

    std::printf("Peer: session disconnected cleanly.\n");

    dp->delete_contained_entities();
    factory->delete_participant(dp);
    return 0;
}

File: integration-tests/java/delete-contained-entities/Peer.java
// integration-tests/java/delete-contained-entities -- peer. Direct Java
// port of c/delete-contained-entities/src/peer.c -- see that file's header
// comment for the full scenario rationale and
// docs/design/integration-test-tier.md for the scenario spec.
//
// Required stdout markers: "Create topic:" x3, "Create writer for topic:",
// "Create reader for topic:" x2, "Peer: session disconnected cleanly."
// Any failure path prints a line starting "FAIL:" and exits nonzero.

import io.zzdds.dcps.Dcps;
import io.zzdds.ext.Zzdds;

import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

public class Peer {
    static final int SAMPLE_COUNT = 5;
    static final int RECEIVE_TARGET = 2;
    static final int READER_READY_TIMEOUT_MS = 10000;
    static final int RECEIVE_TIMEOUT_MS = 20000;
    static final int DISCONNECT_TIMEOUT_MS = 30000;
    static final int POLL_PERIOD_MS = 20;

    static class MatchState {
        final AtomicBoolean everMatched = new AtomicBoolean(false);
        final AtomicInteger matchedCurrentCount = new AtomicInteger(0);
        final AtomicBoolean readerReady = new AtomicBoolean(false);
    }

    static int parseDomain(String[] args) {
        for (int i = 0; i < args.length - 1; i++) {
            if (args[i].equals("-d") || args[i].equals("--domain")) {
                return Integer.parseInt(args[i + 1]);
            }
        }
        return 0;
    }

    static Zzdds.zzdds.DataWriterListenerEx makeWriterListener(MatchState state) {
        return new Zzdds.zzdds.DataWriterListenerEx() {
            public void on_offered_deadline_missed(Dcps.DDS.DataWriter w, Dcps.DDS.OfferedDeadlineMissedStatus s) {}
            public void on_offered_incompatible_qos(Dcps.DDS.DataWriter w, Dcps.DDS.OfferedIncompatibleQosStatus s) {}
            public void on_liveliness_lost(Dcps.DDS.DataWriter w, Dcps.DDS.LivelinessLostStatus s) {}

            public void on_publication_matched(Dcps.DDS.DataWriter w, Dcps.DDS.PublicationMatchedStatus s) {
                state.matchedCurrentCount.set(s.get_current_count());
                if (s.get_current_count() > 0) state.everMatched.set(true);
            }

            public void on_reliable_reader_ready(int readerHandle, boolean isReady) {
                if (isReady) state.readerReady.set(true);
            }
        };
    }

    static Dcps.DDS.DataReaderListener makeReaderListener(MatchState state) {
        return new Dcps.DDS.DataReaderListener() {
            public void on_requested_deadline_missed(Dcps.DDS.DataReader r, Dcps.DDS.RequestedDeadlineMissedStatus s) {}
            public void on_requested_incompatible_qos(Dcps.DDS.DataReader r, Dcps.DDS.RequestedIncompatibleQosStatus s) {}
            public void on_sample_rejected(Dcps.DDS.DataReader r, Dcps.DDS.SampleRejectedStatus s) {}
            public void on_liveliness_changed(Dcps.DDS.DataReader r, Dcps.DDS.LivelinessChangedStatus s) {}
            public void on_data_available(Dcps.DDS.DataReader r) {}
            public void on_sample_lost(Dcps.DDS.DataReader r, Dcps.DDS.SampleLostStatus s) {}

            public void on_subscription_matched(Dcps.DDS.DataReader r, Dcps.DDS.SubscriptionMatchedStatus s) {
                state.matchedCurrentCount.set(s.get_current_count());
                if (s.get_current_count() > 0) state.everMatched.set(true);
            }
        };
    }

    public static void main(String[] args) throws Exception {
        int domainId = parseDomain(args);

        Dcps.DDS.DomainParticipantFactory factory =
            (Dcps.DDS.DomainParticipantFactory) io.zzdds.runtime.ZzddsRuntime.createFactory();
        if (factory == null) {
            System.err.println("FAIL: createFactory() failed");
            System.exit(1);
        }

        Dcps.DDS.DomainParticipant dp = factory.create_participant(domainId, null, null, 0);
        if (dp == null) {
            System.err.println("FAIL: create_participant() failed on domain " + domainId);
            System.exit(1);
        }

        if (SessionEventTypeSupport.register(dp, "SessionEvent") != 0) {
            System.err.println("FAIL: register SessionEventTypeSupport failed");
            System.exit(1);
        }

        Dcps.DDS.Topic out1Topic = dp.create_topic("SessionOut1", "SessionEvent", null, null, 0);
        Dcps.DDS.Topic out2Topic = dp.create_topic("SessionOut2", "SessionEvent", null, null, 0);
        Dcps.DDS.Topic inTopic = dp.create_topic("SessionIn", "SessionEvent", null, null, 0);
        if (out1Topic == null || out2Topic == null || inTopic == null) {
            System.err.println("FAIL: create_topic() failed");
            System.exit(1);
        }
        System.out.println("Create topic: SessionOut1");
        System.out.println("Create topic: SessionOut2");
        System.out.println("Create topic: SessionIn");

        Dcps.DDS.Publisher pub = dp.create_publisher(null, null, 0);
        Dcps.DDS.Subscriber sub = dp.create_subscriber(null, null, 0);
        if (pub == null || sub == null) {
            System.err.println("FAIL: create_publisher/create_subscriber failed");
            System.exit(1);
        }

        Dcps.DDS.DataWriterQos dwQos = new Dcps.DDS.DataWriterQos();
        pub.get_default_datawriter_qos(dwQos);
        dwQos.get_reliability().set_kind(Dcps.DDS.ReliabilityQosPolicyKind.RELIABLE_RELIABILITY_QOS);
        dwQos.get_history().set_kind(Dcps.DDS.HistoryQosPolicyKind.KEEP_ALL_HISTORY_QOS);

        Dcps.DDS.DataWriter inDw = pub.create_datawriter(inTopic, dwQos, null, 0);
        if (inDw == null) {
            System.err.println("FAIL: create_datawriter(SessionIn) failed");
            System.exit(1);
        }
        System.out.println("Create writer for topic: SessionIn");

        Dcps.DDS.DataReaderQos drQos = new Dcps.DDS.DataReaderQos();
        sub.get_default_datareader_qos(drQos);
        drQos.get_reliability().set_kind(Dcps.DDS.ReliabilityQosPolicyKind.RELIABLE_RELIABILITY_QOS);
        drQos.get_history().set_kind(Dcps.DDS.HistoryQosPolicyKind.KEEP_ALL_HISTORY_QOS);

        Dcps.DDS.DataReader out1Dr = sub.create_datareader(out1Topic, drQos, null, 0);
        Dcps.DDS.DataReader out2Dr = sub.create_datareader(out2Topic, drQos, null, 0);
        if (out1Dr == null || out2Dr == null) {
            System.err.println("FAIL: create_datareader() failed");
            System.exit(1);
        }
        System.out.println("Create reader for topic: SessionOut1");
        System.out.println("Create reader for topic: SessionOut2");

        MatchState writerState = new MatchState();
        MatchState out1State = new MatchState();
        MatchState out2State = new MatchState();

        Zzdds.zzdds.DataWriter zInDw = (Zzdds.zzdds.DataWriter) io.zzdds.runtime.ZzddsRuntime.asZzddsDataWriter(inDw);
        if (zInDw == null) {
            System.err.println("FAIL: asZzddsDataWriter() failed");
            System.exit(1);
        }
        if (zInDw.set_listener_ex(makeWriterListener(writerState), Dcps.DDS.PUBLICATION_MATCHED_STATUS.value) != 0) {
            System.err.println("FAIL: set_listener_ex (writer) failed");
            System.exit(1);
        }
        if (out1Dr.set_listener(makeReaderListener(out1State), Dcps.DDS.SUBSCRIPTION_MATCHED_STATUS.value) != 0 ||
            out2Dr.set_listener(makeReaderListener(out2State), Dcps.DDS.SUBSCRIPTION_MATCHED_STATUS.value) != 0)
        {
            System.err.println("FAIL: set_listener (reader) failed");
            System.exit(1);
        }

        SessionEventDataWriter inWriter = new SessionEventDataWriter(inDw);
        SessionEventDataReader out1Reader = new SessionEventDataReader(out1Dr);
        SessionEventDataReader out2Reader = new SessionEventDataReader(out2Dr);

        // Gate on the session's reader actually being registered, not just
        // matched -- see Session.java's matching comment for why.
        long deadline = System.currentTimeMillis() + READER_READY_TIMEOUT_MS;
        while (!writerState.readerReady.get()) {
            if (System.currentTimeMillis() > deadline) {
                System.err.println("FAIL: no reliable reader became ready within " + (READER_READY_TIMEOUT_MS / 1000) + "s");
                System.exit(1);
            }
            Thread.sleep(POLL_PERIOD_MS);
        }

        for (int seq = 0; seq < SAMPLE_COUNT; seq++) {
            Session_event.SessionEvent ev = new Session_event.SessionEvent();
            ev.set_seq(seq);
            if (inWriter.write(ev, 0L) != 0) {
                System.err.println("FAIL: write() failed at seq=" + seq);
                System.exit(1);
            }
        }
        System.out.println("Peer: wrote " + SAMPLE_COUNT + " samples on SessionIn");

        int received1 = 0, received2 = 0;
        long waitDeadline = System.currentTimeMillis() + RECEIVE_TIMEOUT_MS;
        while (received1 < RECEIVE_TARGET || received2 < RECEIVE_TARGET) {
            if (System.currentTimeMillis() > waitDeadline) {
                System.err.println("FAIL: peer did not receive from session within " + (RECEIVE_TIMEOUT_MS / 1000)
                    + "s (out1=" + received1 + " out2=" + received2 + ")");
                System.exit(1);
            }
            SessionEventDataReader.Sample[] s1 = out1Reader.take_n(
                SAMPLE_COUNT, Dcps.DDS.ANY_SAMPLE_STATE.value, Dcps.DDS.ANY_VIEW_STATE.value, Dcps.DDS.ANY_INSTANCE_STATE.value);
            for (SessionEventDataReader.Sample s : s1) if (s.validData) received1++;
            SessionEventDataReader.Sample[] s2 = out2Reader.take_n(
                SAMPLE_COUNT, Dcps.DDS.ANY_SAMPLE_STATE.value, Dcps.DDS.ANY_VIEW_STATE.value, Dcps.DDS.ANY_INSTANCE_STATE.value);
            for (SessionEventDataReader.Sample s : s2) if (s.validData) received2++;
            if (received1 < RECEIVE_TARGET || received2 < RECEIVE_TARGET) Thread.sleep(POLL_PERIOD_MS);
        }
        System.out.println("Peer: received from session (out1=" + received1 + " out2=" + received2 + ").");

        deadline = System.currentTimeMillis() + DISCONNECT_TIMEOUT_MS;
        while (!(writerState.everMatched.get() && writerState.matchedCurrentCount.get() == 0 &&
                 out1State.everMatched.get() && out1State.matchedCurrentCount.get() == 0 &&
                 out2State.everMatched.get() && out2State.matchedCurrentCount.get() == 0)) {
            if (System.currentTimeMillis() > deadline) {
                System.err.println("FAIL: peer never saw a clean disconnect within " + (DISCONNECT_TIMEOUT_MS / 1000) + "s");
                System.exit(1);
            }
            Thread.sleep(POLL_PERIOD_MS);
        }

        System.out.println("Peer: session disconnected cleanly.");

        dp.delete_contained_entities();
        factory.delete_participant(dp);
    }
}
